### PR TITLE
feat: Add config serialization and loading support

### PR DESCRIPTION
**Pull request by rahul-tuli**
# PR 1: Add Config Serialization and Loading Support

## Summary
Implements `from_pretrained`, `to_dict`, and `from_dict` methods for SpeculatorModelConfig to enable loading and saving configurations from the HuggingFace Hub and local paths. This is a prerequisite for model implementations that need these serialization capabilities.

### Core Implementation
- **SpeculatorModelConfig.from_pretrained()**: Load configs from HF Hub or local paths with automatic subclass detection
- **SpeculatorModelConfig.from_dict()**: Polymorphic instantiation from dictionaries using registry pattern
- **EagleSpeculatorConfig.to_dict()**: Proper serialization of all config attributes
- **EagleSpeculatorConfig.from_dict()**: Handle nested config reconstruction (transformer_layer_config, etc.)

#### Registry Pattern
```python
# Base class uses registry for polymorphic instantiation
model_type = config_dict.get("speculators_model_type")
config_class = cls.registry.get(model_type)
return config_class(**config_dict)
```

#### Nested Config Handling
- Automatically converts dict representations to proper config objects:
  - `transformer_layer_config` → LlamaConfig
  - `speculators_config` → SpeculatorsConfig  
  - `token_proposals` → TokenProposalConfig

#### Comment Improvements
- Updated comments to explain "why" rather than "what"
- Removed redundant documentation

### Testing
- **test_config_loading.py**: Comprehensive unit tests for serialization roundtrips
- **Test Coverage**:
  - to_dict/from_dict roundtrip preservation
  - from_pretrained with architecture inference
  - Nested config handling
  - Extra kwargs override behavior

### API Changes
- No breaking changes - only adds new functionality
- Maintains compatibility with HuggingFace PretrainedConfig

## Example Usage
```python
# Load from HuggingFace Hub
config = SpeculatorModelConfig.from_pretrained("org/model-name")

# Load from local path
config = SpeculatorModelConfig.from_pretrained("/path/to/model")

# Serialize and deserialize
config_dict = config.to_dict()
new_config = SpeculatorModelConfig.from_dict(config_dict)
```

## Review Notes
- The registry pattern allows future speculator types to be added without modifying base class
- All changes maintain backward compatibility
- Test coverage includes edge cases like missing model_type fields

### CODE REVIEW

File: src/speculators/config.py [hunk 1]
```diff
@@ -238,7 +238,131 @@ def from_pretrained(
         :param kwargs: Additional keyword arguments to pass to the config.
         :return: A SpeculatorModelConfig object with the loaded parameters.
         """
-        raise NotImplementedError("from_pretrained is not implemented yet.")
+        # Ensure registry is populated with all available speculator configs
+        cls.auto_populate_registry()
```
Comment (brian-dellabetta) [inline src/speculators/config.py]: is this an established pattern in pydantic? i can't find any documentation on it. 

if we are modifying the class definition, it seems like a weird side effect to have in a `@classmethod`?

**Review by brian-dellabetta — APPROVED** [review]

File: src/speculators/config.py (hunk 1 above)
Comment (markurtz) [reply src/speculators/config.py]: auto_populate_registry is a utility added on top of Pydantic within the utils of this repo -- found in the PydanticClassRegistryMixin extension classes. We definitely should not need to call it here, though, as the auto populate will be called on import on the module automatically. Going to remove this line alongside the finalization of this diff work I'm helping work through

**Review by markurtz — CHANGES_REQUESTED** [review]
Noting here, working through simplifying this diff

**Review by markurtz — APPROVED** [review]
Removing my request changes block now that it's been reworked to utilize Pydantic builtins and reduce the complexity of the logic

File: tests/integration/test_config.py [hunk 2]
```diff
@@ -41,11 +40,8 @@ def test_verifier_config_from_verifier_config():
 
 @pytest.mark.smoke
 def test_speculator_model_config_from_pretrained():
-    # swap for real config once implemented
-    with pytest.raises(NotImplementedError) as exc_info:
-        SpeculatorModelConfig.from_pretrained("test/model")
-
-    assert "from_pretrained is not implemented yet" in str(exc_info.value)
+    # Implement loading once real config is available
+    assert True
```
Comment (brian-dellabetta) [inline tests/integration/test_config.py:44]: If we are only asserting True, we should remove this test?

**Review by brian-dellabetta — APPROVED** [review]
nit to remove extraneous test, otherwise LGTM

File: tests/integration/test_config.py (hunk 2 above)
Comment (markurtz) [reply tests/integration/test_config.py:44]: Fair to bring up, going to keep it in just to ensure we don't miss it in #39 diff that will be landing very soon after
